Checkout on Otto

vkaracic opened the pull request:
Introducing a new CommerceConfiguration. When it's enabled all the checkout functionality will be forwarded to Otto. 

@mattdrayer @clintonb can you review this please?
@mjfrey

edx-webhook commented:
Thanks for the pull request, @vkaracic! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request.

To automatically create an OSPR issue for this pull request, just visit this link: https://openedx-webhooks.herokuapp.com/github/process_pr?number=11503&repo=edx%2Fedx-platform

mattdrayer commented on common/djangoapps/course_modes/views.py:
```diff
@@ -39,7 +38,7 @@ class ChooseModeView(View):
 
     @method_decorator(transaction.non_atomic_requests)
-    def dispatch(self, *args, **kwargs):        # pylint: disable=missing-docstring
+    def dispatch(self, *args, **kwargs):  # pylint: disable=missing-docstring
```
Would it be possible to add a docstring for this method and remove the pylint override?

vkaracic replied:
I'm not completely sure about all the use-cases for this view so I'm going to place a general description for this code.

mattdrayer commented on common/djangoapps/student/views.py:
```diff
@@ -735,6 +736,8 @@ def dashboard(request):
         'course_programs': course_programs,
         'disable_courseware_js': True,
         'xseries_credentials': xseries_credentials,
+        'use_ecommerce_payment_flow': EcommerceService().is_enabled(),
```
I noticed that in course_modes.views.py above the 'use_ecommerce_payment_flow' parameter is conditionally available based on sku, versus in this view the parameter always available.  Does it make sense for them to both work the same way?  This also goes for 'ecommerce_payment_page'

mattdrayer replied:
Better to set to a variable and reference that twice, versus building up the object twice.

vkaracic replied:
The problem is that the ``course_modes`` view is for one course and here it is for all the courses in the dashboard, therefor I’ve placed a condition in the template (which is the only one that iterates through the courses) to check if the mode has a SKU as well.

mattdrayer commented on lms/djangoapps/commerce/models.py:
```diff
@@ -1,3 +1,19 @@
+
+class CommerceConfiguration(ConfigurationModel):
+    """ Commerce configuration """
+
```
I know I've asked about this before, but I just want to ask again if this is the right location/time to add a foreign key back to the Site.

mattdrayer replied:
And if so, maybe the name of this model should be SiteCommerceConfiguration.

vkaracic replied:
Since this is not my code, I'd like to include @clintonb to weigh in.

clintonb replied:
I may have initially wrote this code as a prototype; but, if you are submitting the PR @vkaracic, it's your code.

That being said, yes, this would be the place to add multi-tenancy support. That should be tested extensively (obviously).

mattdrayer commented on lms/djangoapps/commerce/utils.py:
```diff
@@ -32,3 +37,19 @@ def audit_log(name, **kwargs):
+
+    def is_enabled(self):
+        """ Check if the service is enabled and that it's not a microsite. """
+        return self.config.checkout_on_ecommerce_service and not microsite.is_request_in_microsite()
```
It would be better to use theming.helpers.is_request_in_themed_site() here

vkaracic replied:
Is that a functionality that will be used in the future for microsites? Because it's practically the same thing: https://github.com/edx/edx-platform/blob/master/openedx/core/djangoapps/theming/helpers.py#L34

mattdrayer replied:
Yes, we will be eliminating "microsite" from the platform over the next 2-3 weeks, replacing it with theming and the Django Sites framework.  So any code references to microsite should be instead directed through theming.

vkaracic replied:
OK, awesome, I'll do the changes right away.

mattdrayer commented on lms/djangoapps/commerce/tests/test_utils.py:
```diff
@@ -16,3 +17,33 @@ def test_log_message(self, mock_log):
+
+    def test_is_enabled_for_microsites(self):
+        """Verify that is_enabled() returns false in case of a microsite."""
+        patcher = patch('microsite_configuration.microsite.is_request_in_microsite')
```
It would be better to use is_request_in_themed_site() here

mattdrayer commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -232,6 +234,75 @@ def test_course_about_in_cart(self):
+        self.assertEqual(response.status_code, 200)
+        self.assertIn(ecommerce_enrollment_link, response.content)
+
+        # Test for logged in user
```
Please move this use case into its own test

mattdrayer commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -232,6 +234,75 @@ def test_course_about_in_cart(self):
+        self.assertEqual(response.status_code, 200)
+        self.assertIn(ecommerce_enrollment_link, response.content)
+
+        # Test for logged in user
```
Please move this use case into its own test

mattdrayer commented on lms/djangoapps/courseware/views.py:
```diff
@@ -855,6 +856,7 @@ def course_about(request, course_id):
         course = get_course_with_access(request.user, permission, course_key)
+        modes = CourseMode.modes_for_course_dict(course_key)
 
         if microsite.get_value('ENABLE_MKTG_SITE', settings.FEATURES.get('ENABLE_MKTG_SITE', False)):
```
Not really part of this PR but would be nice if we could change this to use theming.helpers.get_value() instead of referencing microsite directly

mattdrayer commented on common/djangoapps/course_modes/views.py:
```diff
@@ -129,15 +129,21 @@ def get(self, request, course_id, error=None):
+            context["verified_description"] = verified_mode.description
+
+            if verified_mode.sku:
+                context["use_ecommerce_payment_flow"] = EcommerceService().is_enabled()
```
Better to set to a variable and reference that twice, versus building up the object twice.

mattdrayer commented on lms/djangoapps/courseware/views.py:
```diff
@@ -894,6 +895,18 @@ def course_about(request, course_id):
+        mode = ''
+        ecomm_service = EcommerceService()
+        if ecomm_service.is_enabled() and (
+            'professional' in modes or 'no-id-professional' in modes
```
It drives me crazy to see these specific mode strings in the platform code -- I know they are everywhere, and I don't have a good solution at the moment, but there has to be a better way (enum?) to make the system more resilient.

mattdrayer commented on lms/djangoapps/courseware/views.py:
```diff
@@ -894,6 +895,18 @@ def course_about(request, course_id):
+            'professional' in modes or 'no-id-professional' in modes
+        ):
+            mode = modes.get('professional', '') or modes.get('no-id-professional', '')
+            ecommerce_checkout_link = "{}?sku={}" .format(ecomm_service.payment_page_url(), mode.sku)
```
Should the EcommerceService own the construction of the checkout link, versus having this logic in the view?  Then it could be used in more places.

mattdrayer commented on lms/djangoapps/courseware/views.py:
```diff
@@ -894,6 +895,18 @@ def course_about(request, course_id):
+            mode = modes.get('professional', '') or modes.get('no-id-professional', '')
+            ecommerce_checkout_link = "{}?sku={}" .format(ecomm_service.payment_page_url(), mode.sku)
+
+            reg_then_add_to_cart_link = "{}?course_id={}&enrollment_action=add_to_ecomm_cart&checkout_url={}".format(
```
Another possible item for the EcommerceService to own/manage.

mattdrayer commented on lms/djangoapps/courseware/views.py:
```diff
@@ -1581,8 +1597,7 @@ def financial_assistance_form(request):
             mode_slug=CourseMode.VERIFIED
-        ).exists()
-        and enrollment.mode != CourseMode.VERIFIED
+        ).exists() and enrollment.mode != CourseMode.VERIFIED
```
I'm not sure if this makes any sense from a performance standpoint but would moving the enrollment.mode check to the first position in the conditional prevent the database query from taking place if enrollment.mode does not equal CourseMode.VERIFIED?

vkaracic replied:
I just checked with the good folks at #python IRC, they said that if the first condition is False the second one won't be executed, therefor I'm moving this to the first position. Nice catch!

mattdrayer commented:
@vkaracic done with my first pass -- let me know when you're ready for another.

mattdrayer commented:
Also it looks like you have a quality violation to address

vkaracic commented:
@mattdrayer thanks for doing the review! I've addressed most of the comments, the ones with theming methods are left (I have a question for it: https://github.com/edx/edx-platform/pull/11503#discussion_r52730366) and the ``CommerceConfiguration`` comments, which I'm waiting for @clintonb to weigh in.

vkaracic commented:
@mattdrayer I believe I have addressed the 'microsite2theming' comments with this commit: https://github.com/edx/edx-platform/commit/e0b2d3f20092757434c7274a36937877328a4493

Aside from the ``CommerceConfiguration`` comments I believe the PR is ready for another look.

mattdrayer commented:
LGTM @vkaracic!  :+1:  aside from the open issue regarding Site tie-in which we are looking to @clintonb for some advice

vkaracic commented:
@bderusha this is the PR that I've mentioned. Please take a look.

clintonb commented on lms/djangoapps/commerce/tests/test_utils.py:
```diff
@@ -16,3 +18,54 @@ def test_log_message(self, mock_log):
+        super(EcommerceServiceTests, self).setUp()
+
+    def test_is_enabled(self):
+        """Verify that is_enabled() returns True."""
```
What about when it is disabled?

clintonb replied:
Seeing your other tests, this docstring needs to be more informative: Verify `is_enabled() returns True when not used for a microsite.

clintonb commented on lms/djangoapps/commerce/tests/test_utils.py:
```diff
@@ -16,3 +18,54 @@ def test_log_message(self, mock_log):
+
+    def test_is_enabled_for_microsites(self):
+        """Verify that is_enabled() returns false in case of a microsite."""
+        patcher = patch('openedx.core.djangoapps.theming.helpers.is_request_in_themed_site')
```
Why not use the decorator? Also, should this be setting a return value?

clintonb commented on lms/djangoapps/commerce/tests/test_utils.py:
```diff
@@ -16,3 +18,54 @@ def test_log_message(self, mock_log):
+        """Verify that is_enabled() returns false in case of a microsite."""
+        patcher = patch('openedx.core.djangoapps.theming.helpers.is_request_in_themed_site')
+        patcher.start()
+        is_enabled = EcommerceService().is_enabled()
```
This doesn't need to be a variable.

vkaracic replied:
You mean I should move this to the assert argument?
``assertFalse(EcommerceService().is_enabled())``

clintonb commented on lms/djangoapps/commerce/tests/test_utils.py:
```diff
@@ -16,3 +18,54 @@ def test_log_message(self, mock_log):
+        self.assertFalse(is_enabled)
+        patcher.stop()
+
+    @patch('django.conf.settings.ECOMMERCE_PUBLIC_URL_ROOT', 'http://ecommerce_url')
```
Use `@override_settings` here and everywhere else you are overriding settings.

clintonb commented on lms/djangoapps/commerce/utils.py:
```diff
@@ -32,3 +39,29 @@ def audit_log(name, **kwargs):
+        return self.config.checkout_on_ecommerce_service and not helpers.is_request_in_themed_site()
+
+    def payment_page_url(self):
+        """ Return the URL for the checkout page. Example:
```
Move the example to it's own section:

```python
Return the URL for the checkout page.

Example:
  http://localhost:8002/basket/single_item/
```

clintonb commented on lms/djangoapps/commerce/models.py:
```diff
@@ -1,3 +1,25 @@
 """
-This file is intentionally empty. Django 1.6 and below require a models.py file for all apps.
+Configuration model for ecommerce service (Otto).
```
This docstring describes the class below. The module-level docstring should describe the module. It can be as simple as "Commerce-related models."

clintonb commented on lms/djangoapps/commerce/admin.py:
```diff
@@ -0,0 +1,8 @@
+from commerce.models import CommerceConfiguration
+from config_models.admin import ConfigurationModelAdmin
+
+admin.site.register(CommerceConfiguration, ConfigurationModelAdmin)
```
Extend this and update the list view to display the current values. Our current list views are not useful. Bonus points if you can update `ConfigurationModelAdmin` to make this happen for all configurations.

vkaracic replied:
Can you please be more specific which values should be displayed in the admin panel. This is what I currently see: http://i.imgur.com/exUJHyr.png

clintonb replied:
In that case, you're all set.

clintonb commented on lms/djangoapps/commerce/utils.py:
```diff
@@ -32,3 +39,29 @@ def audit_log(name, **kwargs):
+
+    def register_then_add_to_cart_path(self, course_id, sku):
+        """ Construct the path for a user to register or log in and redirect to the ecommerce checkout page. """
+        return "{}?course_id={}&enrollment_action=add_to_ecomm_cart&checkout_url={}".format(
```
Use [`urllib.urlencode`](https://docs.python.org/2/library/urllib.html#urllib.urlencode).

clintonb commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -232,6 +234,71 @@ def test_course_about_in_cart(self):
 
+    def prepare_course_about_with_ecommerce(self, min_price=0):
+        """ Helper function to prepare an commerce configuration, course and request for tests. """
+        # Create an ecommerce configuration
```
Unnecessary comment is unnecessary.

clintonb commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -232,6 +234,71 @@ def test_course_about_in_cart(self):
+            checkout_on_ecommerce_service=True,
+            single_course_checkout_page=checkout_page
+        )
+        course = CourseFactory.create(org="new", number="unenrolled", display_name="course")
```
Single quotes or double quotes. Pick one.

clintonb commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -232,6 +234,71 @@ def test_course_about_in_cart(self):
+        checkout_page, sku, course, request = self.prepare_course_about_with_ecommerce()
+        request.user = AnonymousUser()
+        mako_middleware_process_request(request)
+        ecommerce_enrollment_link = (
```
There is a lot of duplicated code across these tests. Combine it into an assertion:

```python
def assert_enrollment_link_present(self, request, href):
  ...
```

bderusha replied:
Additionally I could see these 4 tests using DDT

```python
@ddt.data(
    (None, True, '<a href="/register?course_id={}&enrollment_action=add_to_ecomm_cart&checkout_url={}?sku={}" class="add-to-cart">'),
    ...etc
)
def test_ecommerce_checkout_with_anonymous(self, min_price, is_anonymous, enrollment_link_template):
...
```

vkaracic replied:
I ended up splitting the tests into two with DDT, because two of them test when the shopping cart setting is enabled: 559b964d9669b43b9ed4905b2e3ebf4cc77882d1

clintonb commented on lms/djangoapps/courseware/views.py:
```diff
@@ -892,7 +895,18 @@ def course_about(request, course_id):
+                reg_url=reverse('register_user'), course_id=urllib.quote(str(course_id))
+            )
+
+        ecommerce_checkout_link = ''
```
Please add comments explaining what is going on here. I'm not sure what you're doing or why.

vkaracic replied:
Here I'm assigning values to ``mode`` and ``ecommerce_checkout_link`` because otherwise they are only assigned in the below if-block and referencing them in the ``render_to_response`` call would result in an error if they're not assigned.

The ``course_about`` view shows the about page of a course and button to enroll, or add the course to the shopping cart if it's a (no_id)professional one. If the ecommerce checkout service is enabled (and it's a (no_id)professional course) new links are constructed for that course and SKU that point the 'add to cart' button to Otto instead of the old shopping cart.

clintonb replied:
The comments should be in the code.

clintonb commented on lms/templates/courseware/course_about.html:
```diff
@@ -152,14 +153,34 @@
             reg_element_id = "reg_then_add_to_cart"
           %>
+          <% if ecommerce_checkout:
+              if user.is_authenticated():
```
There should be no need to redirect to login/registration manually. Make sure the checkout view (both in LMS and Otto) requires login. Django will handle the rest.

clintonb commented:
:+1: pending:

- [ ] All comments acknowledged
- [ ] Comments added around Professional Education changes (see comment)
- [ ] Manual redirects to login/registration page removed

bderusha commented on common/djangoapps/course_modes/views.py:
```diff
@@ -117,7 +123,8 @@ def get(self, request, course_id, error=None):
 
         context = {
-            "course_modes_choose_url": reverse("course_modes_choose", kwargs={'course_id': course_key.to_deprecated_string()}),
+            "course_modes_choose_url": reverse("course_modes_choose",
```
NIT: put both arguments one singly indented lines.

```python
reverse(
    "course_modes_choose",
    kwargs={'course_id': course_key.to_deprecated_string()}
),
```

bderusha commented on common/djangoapps/student/views.py:
```diff
@@ -734,9 +735,16 @@ def dashboard(request):
         'course_programs': course_programs,
         'disable_courseware_js': True,
-        'xseries_credentials': xseries_credentials,
+        'xseries_credentials': xseries_credentials
```
Why are we removing this comma?  I thought our convention was to end dicts with commas

bderusha commented on common/djangoapps/student/views.py:
```diff
@@ -734,9 +735,16 @@ def dashboard(request):
+    if ecommerce_service.is_enabled():
+        context.update({
+            'use_ecommerce_payment_flow': True,
+            'ecommerce_payment_page': ecommerce_service.payment_page_url(),
```
like here :)

bderusha commented on lms/djangoapps/commerce/utils.py:
```diff
@@ -32,3 +39,29 @@ def audit_log(name, **kwargs):
+        """ Construct the URL to the ecommerce checkout page and include a product. """
+        return "{}?sku={}".format(self.payment_page_url(), sku)
+
+    def register_then_add_to_cart_path(self, course_id, sku):
```
Can we be consistent about using `url` or `path` in method names, unless there is a subtle distinction that I am missing.  I prefer `url` personally, but either is okay as long as it's consistent.

vkaracic replied:
I like to use ``URL`` for 'full' links that include the protocol: ``https://www.example.com/bla/blah`` while ``path`` would be for paths like ``/path/to/something/`` if that makes sense?

bderusha replied:
Gotcha.  Keep it! :100:

bderusha commented on lms/djangoapps/courseware/views.py:
```diff
@@ -1577,12 +1594,11 @@ def financial_assistance_form(request):
         {'name': enrollment.course_overview.display_name, 'value': unicode(enrollment.course_id)}
         for enrollment in CourseEnrollment.enrollments_for_user(user).order_by('-created')
-        if CourseMode.objects.filter(
+        if enrollment.mode != CourseMode.VERIFIED and CourseMode.objects.filter(
```
NIT:  rather than having a full query for an if statement this would be more readable if it were pulled out into a variable 

``` python
is_non_verified_enrollment = enrollment.mode != CourseMode.VERIFIED
course_has_unexpired_verified_course_mode = CourseMode.objects.filter(...)

if is_non_verified_enrollment and course_has_unexpired_verified_course_mode:
   ...
```

vkaracic replied:
It would be more readable, but this way we could avoid a database query if the first condition is not satisfied.

bderusha replied:
Ah good point.  Maybe a comment then?  It's just a little bit much to look at and quickly understand what this is doing.

bderusha commented on lms/static/js/student_account/views/FinishAuthView.js:
```diff
@@ -146,6 +148,8 @@
                     */
                     this.updateTaskDescription(gettext("Adding the selected course to your cart"));
                     shoppingCartInterface.addCourseToCart( this.courseId );
+                } else if ( this.enrollmentAction === 'add_to_ecomm_cart' && this.courseId) {
```
ASIDE: The more of these we add the more this section looks like it could be a little cleaner with a switch statement on `enrollmentAction`

bderusha commented on themes/edx.org/lms/templates/course_modes/choose.html:
```diff
@@ -21,7 +21,7 @@
                 title.attr("aria-expanded", "false");
             }
-        }
+        };
```
Can we make this change in both places? https://github.com/edx/edx-platform/pull/11503/files#diff-81310268a943607d9094f376f1c1e863R24

bderusha commented:
:+1: pending:

- [ ] All non-NIT/ASIDE comments acknowledged (though fixing those while you're at it would be :rocket: )
- [ ] Fix failing tests
- [ ] Squash commits

clintonb commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -232,6 +233,45 @@ def test_course_about_in_cart(self):
         self.assertEqual(response.status_code, 200)
         self.assertIn(in_cart_span, response.content)
 
+    def assert_enrollment_link_present(self, is_anonymous, _id=False):
```
Document the kwargs.

clintonb commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -232,6 +233,45 @@ def test_course_about_in_cart(self):
         self.assertIn(in_cart_span, response.content)
 
+    def assert_enrollment_link_present(self, is_anonymous, _id=False):
+        """ Prepare ecommerce checkout data and assert if the ecommerce link is contained in the response. """
```
Document the kwargs.

clintonb commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -232,6 +233,45 @@ def test_course_about_in_cart(self):
+            checkout_on_ecommerce_service=True,
+            single_course_checkout_page=checkout_page
+        )
+        course = CourseFactory.create(org='new', number='unenrolled', display_name='course')
```
Nit: If you don't specify these kwargs, they will be automatically generated for you. Given that we don't really care what they are, this code would be cleaner if you removed these kwargs.

clintonb commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -232,6 +233,45 @@ def test_course_about_in_cart(self):
+        course = CourseFactory.create(org='new', number='unenrolled', display_name='course')
+        CourseModeFactory(mode_slug=CourseMode.PROFESSIONAL, course_id=course.id, sku=sku, min_price=1)
+
+        request = self.request_factory.get(reverse('about_course', args=[course.id.to_deprecated_string()]))
```
`to_deprecated_string()` is itself deprecated. Use `unicode(course.id)`.

vkaracic replied:
In that case I've change all the instances of ``to_deprecated_string()`` in this file.

clintonb commented on common/djangoapps/course_modes/views.py:
```diff
@@ -117,7 +123,10 @@ def get(self, request, course_id, error=None):
-            "course_modes_choose_url": reverse("course_modes_choose", kwargs={'course_id': course_key.to_deprecated_string()}),
+            "course_modes_choose_url": reverse(
+                "course_modes_choose",
+                kwargs={'course_id': course_key.to_deprecated_string()}
```
Use `unicode(course_key)`.

clintonb commented on lms/djangoapps/courseware/tests/test_views.py:
```diff
@@ -232,6 +233,45 @@ def test_course_about_in_cart(self):
+            'sku=TEST123',
+            ' id="">' if _id else ">"
+        )
+        response = views.course_about(request, course.id.to_deprecated_string())
```
Use `unicode(course.id)`.

clintonb commented on lms/djangoapps/courseware/views.py:
```diff
@@ -1577,12 +1593,14 @@ def financial_assistance_form(request):
         for enrollment in CourseEnrollment.enrollments_for_user(user).order_by('-created')
-        if CourseMode.objects.filter(
+
+        # Here we are doing a database query as an if condition because if the first condition fails
```
Nit: this doesn't really need to be commented.

vkaracic replied:
Seems like we have a nit conflict :)
https://github.com/edx/edx-platform/pull/11503#discussion_r52821257

vkaracic replied:
Seems like we have a nit conflict :)
https://github.com/edx/edx-platform/pull/11503#discussion_r52821257
@bderusha

vkaracic commented:
@clintonb I believe I have addressed all your comment. You were right about the login, I've removed that code which rendered the JS additions unnecessary so those are gone as well.

@bderusha as I removed the JS changes this aside https://github.com/edx/edx-platform/pull/11503#discussion_r52785405 might not be applicable anymore. Also I tend to agree with Clinton that this comment https://github.com/edx/edx-platform/pull/11503#discussion_r53178474 isn't necessary so I removed it, I hope that's OK.

bderusha commented:
@vkaracic That's fine

clintonb commented:
Cool.
